### DISCUSSION TITLE
A few V3 questions

### DISCUSSION THREAD

**Discussion by PureKrome** [xunit/xunit#3104] [General]
👋🏻 G'Day @bradwilson 

So far so good with v3! I have a few more questions though and I've _tried_ to read the docs

- if i have a solution with multiple tests, what is the recommended way to run all the tests is a single go? `dotnet test`? or is it recommended to `dotnet run --project` per project?
- what is the parallelization situation with v3? parallel by 'default'? e.g. 1 'core' per class? max number of cores used? a single core for all tests in a 'collection' etc?

edit: important point i missed: this is all 💯 targeting **Microsoft Testing Platform**.

**Comment by bradwilson** [comment]
> if i have a solution with multiple tests, what is the recommended way to run all the tests is a single go? 

That's a decision that's up to you. We have always provided multiple ways to run tests so that developers can choose the way they prefer.

If you're asking what I do for a single test assembly during interactive development, I use `dotnet run` (via Ctrl+F5 in Visual Studio) to run all tests, and then rely on Test Explorer (also typically via keyboard shortcuts) to run or debug a single test.

If you're asking what I do for multiple test assemblies during command line builds (both for local verification and CI), then I use `xunit.v3.runner.console` to run multiple assemblies simultaneously.

> what is the parallelization situation with v3?

The defaults have not changed. Within a single test assembly, we the default is to parallelize by test collection, and the default test collection definition is "one test collection per test class". The ways to override these defaults are the same as in v2: via attributes in source code, configuration files, or command line switches. The same goes for using mechanisms to run multiple test assemblies in parallel with one another: configuration files and command line switches can control that behavior.

https://xunit.net/docs/running-tests-in-parallel
https://xunit.net/docs/configuration-files
https://xunit.net/docs/runsettings

Command line switches for third party runners are documented by those authors.

> this is all 💯 targeting Microsoft Testing Platform

This does not change any of the above recommendations/descriptions, other than the command line switches changing between the xUnit.net native command line UX vs. the Microsoft Testing Platform command line UX.

https://xunit.net/docs/getting-started/v3/microsoft-testing-platform

**Comment by PureKrome** [comment]
Thank you @bradwilson !

> If you're asking what I do for multiple test assemblies during command line builds (both for local verification and CI), then I use `xunit.v3.runner.console` to run multiple assemblies simultaneously

Using this, this uses the "older" VSTest platform?

In the future, will you _personally_ start to move off `xunit.v3.runner.console` to run multiple assemblies simultaneously?

Doesn't `xunit.v3.runner.console` use VSTest underneath?

**Reply by bradwilson** [reply]
xunit.v3.runner.console does not use VSTest. It natively understands how to run xUnit.net test assemblies, and can run assemblies linked against v1, v2, or v3 of xUnit.net.

**Reply by bradwilson** [reply]
To be more concise: `xunit.runner.visualstudio` is the only thing that understands VSTest in the xUnit.net suite. Just like `xunit.v3.runner.console`, it knows how to run all of v1, v2, and v3 xUnit.net tests natively (by leveraging `xunit.v3.runner.utility`), except rather than having a command line UX like `xunit.v3.runner.console` or an MSBuild UX like `xunit.v3.runner.msbuild`, it translates that the requests and results into API calls that VSTest can understand.

Unlike VSTest support, support for Microsoft.Testing.Platform is integrated directly into in-process console runner in v3. That is the thing that is executed when you do `dotnet run`, and the code lives in `xunit.v3.runner.inproc.console`. MTP support gets surfaced in two orthogonal forms:

(A) We can swap out the command line UX from xUnit.net with the one from Microsoft.Testing.Platform, so that when you `dotnet run` you get MTP switches and output, rather than xUnit.net's switches and output (though of course we extend MTP to include versions of all of our command line switches, just in a format that is more in line with MTP).

(B) We can detect when we're being launched with a Microsoft.Testing.Platform host (of which there are currently two I'm aware of) and then communicate with it appropriately. This is equivalent to the code that exists in `xunit.runner.visualstudio` to support VSTest, except that it's natively integrated. Whereas VSTest relies upon an "adapter" that can be discovered at runtime, MTP assumes that test projects are executables which can be launched with well-known command line switches to indicate to the test project that it's being hosted. Technically these command line switches are MTP switches, so even if you don't do choose option A above, we still have special recognition of a very small set of MTP command line switches that will let us know that we are being hosted by MTP.

The two hosts I mentioned that I'm currently aware of:

- A "temporary" MSBuild-based host that is hot-swapped in when you run `dotnet test` against an MTP-compatible project (that's also linked against the `Microsoft.Testing.Platform.MSBuild` NuGet package). This basically bypasses the old VSTest-based engine and runs your code in an entirely different way. A downside of this strategy is that the existing `dotnet test` command line switches are not supported by this MSBuild-based host, so you are forced to understand that you need to use MTP command line switches, after `dotnet test [dotnet-test-options] -- [mtp-options]`, and the entire experience is different. My understanding is that this temporary host will be replaced with something more permanent, presumably built into the .NET SDK (like `dotnet test` is today), but I do not know when that is shipping.

- A "preview" Visual Studio-based host that is disabled by default current in RTM versions of VS, which allows Test Explorer to detect what it believes to be MTP-compatible test projects, and rather than using VSTest to talk to them, it uses MTP. You can verify that this works by removing your reference to `xunit.runner.visualstudio` and then trying to run your tests; if they run, you have the preview mode enabled, and if they don't, then you don't. In VS 2022 17.12.3 (current as this writing), you have to check this box to enable MTP mode:

  ![image](https://github.com/user-attachments/assets/9e7fdd6a-1697-4594-9ca4-4caba50851dd)

I assume at some point, other test runners (like VS Code, Resharper/Rider, CodeRush, etc.) will eventually also integrate MTP support into their test runners. Doing so will require them to implement the same host RPC protocol that's in use by Test Explorer today, or else none of the MTP-compatible test projects will understand how to communicate with them. They are already doing similar re-implementations of the VSTest APIs that test projects are expecting to be called through.

The fact that these other runners don't yet support MTP is a reason why it's likely premature to rely solely on MTP and remove VSTest support (meaning, removing `xunit.runner.visualstudio`) unless they can guarantee that 100% of their developers and CI environments are compatible with MTP.

**Reply by bradwilson** [reply]
> We can detect when we're being launched with a Microsoft.Testing.Platform host

If you're curious how this works, we are injecting a `Main` function as the entry point into your test project at build time via MSBuild. We provide a version for both xUnit.net native mode and a version for MTP mode.

When we inject the xUnit.net native version, we sniff looking for the MTP host switches:

https://github.com/xunit/xunit/blob/9853a4b92a41b0129c30a70cd040ea5213a016c6/src/xunit.v3.core/Package/content/EntryPoint-xunit.cs#L6-L12

We have inverse logic in MTP mode, so we can detect when we're being hosted by an xUnit.net native host (like `xunit.v3.runner.console`):

https://github.com/xunit/xunit/blob/9853a4b92a41b0129c30a70cd040ea5213a016c6/src/xunit.v3.core/Package/content/EntryPoint-mtp.cs#L6-L12

**Comment by PureKrome** [comment]
thanks again @bradwilson for the detailed answers - sincerely appreciate it.
